### PR TITLE
hypot template

### PR DESCRIPTION
**Pull request by 9il**
Win64 still fails. Help is needed from someone with Win64.
#### Blocked

DMD [issue](https://issues.dlang.org/show_bug.cgi?id=13548): ~~(workaround)~~
#### ~~Contents~~
1. ~~Remove `FPTemporary` usage in `std.complex`.~~ #2637 (merged)
2. ~~Remove `FPTemporary` usage in `std.numeric`.~~ #2637 (merged)
3. Change `hypot` to `template`. ~~<---------- fails~~ (workaround)
4. ~~Clean up imports in `std.complex`.~~ #2659 (merged)
5. ~~Clean up imports in `std.numeric`.~~ #2658 (merged)
6. ~~Change `isfinite, isinf` to `isFinite, isInfinity` in `std.numeric` and `std.math`.~~ -> #2636 (merged)

### CODE REVIEW

**Comment by 9il** [comment]
PR is unblocked (workaround).

File: std/math.d [hunk 1]
```diff
@@ -522,7 +522,10 @@ real abs(Num)(Num y) @safe pure nothrow @nogc
     assert(abs(71.6Li) == 71.6L);
     assert(abs(-56) == 56);
     assert(abs(2321312L)  == 2321312L);
-    assert(abs(-1+1i) == sqrt(2.0L));
+    assert(abs(creal(-1+1i)) == sqrt(2.0L));
+    cdouble c = -1+1i;
+    double d = 2.0;
+    assert(abs(c) == sqrt(d));
```
Comment (9il) [inline std/math.d]: Fails on Win64. Any ideas?

Comment (yebblies) [inline std/math.d]: That's weird.  How far off are the values when the comparison is done?

Comment (9il) [inline std/math.d]: `assert(abs(c) == sqrt(d), to!string(abs(c) - sqrt(d)));` prints `0`

Comment (yebblies) [inline std/math.d]: Sure, but I expect if you use a debugger to look at the register contents they'll tell a different story.

Comment (9il) [inline std/math.d]: I am on OS X =(

**Comment by 9il** [comment]
Win64 still fails. Needs help from someone with Win64.

**Comment by JakobOvrum** [comment]
@9il, needs rebase for the autotester.

**Comment by DmitryOlshansky** [comment]
Seems to fail on some 32-bit platforms btw. @9il I have Win64 - where to dig?

**Comment by 9il** [comment]
Disassembling this code would help, thought.

```D
double two = 2.0;
auto a = sqrt(two);
```
```D
auto a = sqrt(2.0)
```
```D
cdouble c = -1+1i;
auto a = hypot(c.re, c.im) ;
```
```D
cdouble c = -1+1i;
auto a = abs(c) ;
```

Can you get it and past please?

**Comment by DmitryOlshansky** [comment]
After merging your pull and rebuilding phobos this D code:
```
import std.math;

void foo1(){
    double two = 2.0;
    auto a = sqrt(two);
}

void foo2(){
    auto a = sqrt(2.0);
}

void foo3(){
    cdouble c = -1+1i;
    auto a = hypot(c.re, c.im);
}

void foo4(){
    cdouble c = -1+1i;
    auto a = abs(c) ;
}
```
produces this ASM:
```
.text:0000000000000140 _D6test644foo1FZv proc near             
.text:0000000000000140
.text:0000000000000140 var_20          = qword ptr -20h
.text:0000000000000140 var_8           = qword ptr -8
.text:0000000000000140
.text:0000000000000140                 push    rbp
.text:0000000000000141                 mov     rbp, rsp
.text:0000000000000144                 sub     rsp, 20h
.text:0000000000000148                 movsd   xmm0, cs:_TMP0
.text:0000000000000151                 movsd   [rbp+var_8], xmm0
.text:0000000000000157                 fld     [rbp+var_8]
.text:000000000000015A                 fsqrt
.text:000000000000015C                 fstp    [rbp+var_20]
.text:000000000000015F                 movsd   xmm1, [rbp+var_20]
.text:0000000000000164                 lea     rsp, [rbp+0]
.text:0000000000000168                 pop     rbp
.text:0000000000000169                 retn
.text:0000000000000169 _D6test644foo1FZv endp

.text:0000000000000188 _D6test644foo2FZv proc near             
.text:0000000000000188
.text:0000000000000188 var_10          = qword ptr -10h
.text:0000000000000188
.text:0000000000000188                 push    rbp
.text:0000000000000189                 mov     rbp, rsp
.text:000000000000018C                 sub     rsp, 10h
.text:0000000000000190                 fld     cs:_TMP0
.text:0000000000000196                 fsqrt
.text:0000000000000198                 fstp    [rbp+var_10]
.text:000000000000019B                 movsd   xmm0, [rbp+var_10]
.text:00000000000001A0                 lea     rsp, [rbp+0]
.text:00000000000001A4                 pop     rbp
.text:00000000000001A5                 retn
.text:00000000000001A5 _D6test644foo2FZv endp

.text:00000000000001C0 _D6test644foo3FZv proc near             
.text:00000000000001C0
.text:00000000000001C0 var_10          = qword ptr -10h
.text:00000000000001C0 var_8           = qword ptr -8
.text:00000000000001C0
.text:00000000000001C0                 push    rbp
.text:00000000000001C1                 mov     rbp, rsp
.text:00000000000001C4                 sub     rsp, 10h
.text:00000000000001C8                 fld     cs:_TMP1
.text:00000000000001CE                 fld     cs:dbl_100
.text:00000000000001D4                 fstp    [rbp+var_8]
.text:00000000000001D7                 fstp    [rbp+var_10]
.text:00000000000001DA                 movsd   xmm1, [rbp+var_10]
.text:00000000000001E0                 movsd   xmm0, [rbp+var_8]
.text:00000000000001E6                 sub     rsp, 20h
.text:00000000000001EA                 movq    rdx, xmm1
.text:00000000000001EF                 movq    rcx, xmm0
.text:00000000000001F4                 call    _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd
.text:00000000000001F9                 add     rsp, 20h
.text:00000000000001FD                 lea     rsp, [rbp+0]
.text:0000000000000201                 pop     rbp
.text:0000000000000202                 retn
.text:0000000000000202 _D6test644foo3FZv endp

.text:0000000000000220 _D6test644foo4FZv proc near             ; DATA XREF: .pdata:$pdata$_D6test644foo4FZvo
.text:0000000000000220                                         ; .pdata:000000000000026Co
.text:0000000000000220
.text:0000000000000220 var_20          = qword ptr -20h
.text:0000000000000220 var_18          = qword ptr -18h
.text:0000000000000220 var_10          = qword ptr -10h
.text:0000000000000220 var_8           = qword ptr -8
.text:0000000000000220
.text:0000000000000220                 push    rbp
.text:0000000000000221                 mov     rbp, rsp
.text:0000000000000224                 sub     rsp, 20h
.text:0000000000000228                 fld     cs:_TMP1
.text:000000000000022E                 fld     cs:dbl_100
.text:0000000000000234                 fstp    [rbp+var_18]
.text:0000000000000237                 fstp    [rbp+var_20]
.text:000000000000023A                 fld     [rbp+var_20]
.text:000000000000023D                 fld     [rbp+var_18]
.text:0000000000000240                 fstp    [rbp+var_8]
.text:0000000000000243                 fstp    [rbp+var_10]
.text:0000000000000246                 lea     rcx, [rbp+var_10]
.text:000000000000024A                 sub     rsp, 20h
.text:000000000000024E                 call    _D3std4math10__T3absTrZ3absFNaNbNiNfrZd
.text:0000000000000253                 add     rsp, 20h
.text:0000000000000257                 lea     rsp, [rbp+0]
.text:000000000000025B                 pop     rbp
.text:000000000000025C                 retn
.text:000000000000025C _D6test644foo4FZv endp
```

**Comment by DmitryOlshansky** [comment]
Also I added T.stringof to 562 line in math.d and T happens to be cfloat. I guess it's one of 'em codegen bugs with built-in complex numbers. I really-really wish we remove them for good.

**Comment by 9il** [comment]
Please, can you add asm with `-O -inline -release` ?

**Comment by DmitryOlshansky** [comment]
foo1 and foo2 were optimized to single ret + stack push/pop.
Others are:
```
.text:0000000000000188 _D6test644foo3FZv proc near            
.text:0000000000000188
.text:0000000000000188 var_10          = qword ptr -10h
.text:0000000000000188 var_8           = qword ptr -8
.text:0000000000000188
.text:0000000000000188                 push    rbp
.text:0000000000000189                 mov     rbp, rsp
.text:000000000000018C                 sub     rsp, 10h
.text:0000000000000190                 fld     cs:_TMP1
.text:0000000000000196                 fld     cs:dbl_100
.text:000000000000019C                 fstp    [rbp+var_8]
.text:000000000000019F                 fstp    [rbp+var_10]
.text:00000000000001A2                 movsd   xmm1, cs:_TMP1
.text:00000000000001AB                 movsd   xmm0, [rbp+var_8]
.text:00000000000001B1                 sub     rsp, 20h
.text:00000000000001B5                 movq    rdx, xmm1
.text:00000000000001BA                 movq    rcx, xmm0
.text:00000000000001BF                 call    _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd
.text:00000000000001C4                 add     rsp, 20h
.text:00000000000001C8                 lea     rsp, [rbp+0]
.text:00000000000001CC                 pop     rbp
.text:00000000000001CD                 retn
.text:00000000000001CD _D6test644foo3FZv endp


.text:00000000000001E8 _D6test644foo4FZv proc near             
.text:00000000000001E8
.text:00000000000001E8 var_10          = qword ptr -10h
.text:00000000000001E8 var_8           = qword ptr -8
.text:00000000000001E8
.text:00000000000001E8                 push    rbp
.text:00000000000001E9                 mov     rbp, rsp
.text:00000000000001EC                 sub     rsp, 10h
.text:00000000000001F0                 fld     cs:_TMP1
.text:00000000000001F6                 fld     cs:dbl_100
.text:00000000000001FC                 fstp    [rbp+var_8]
.text:00000000000001FF                 fstp    [rbp+var_10]
.text:0000000000000202                 movsd   xmm1, cs:_TMP1
.text:000000000000020B                 movsd   xmm0, [rbp+var_8]
.text:0000000000000211                 sub     rsp, 20h
.text:0000000000000215                 movq    rdx, xmm1
.text:000000000000021A                 movq    rcx, xmm0
.text:000000000000021F                 call    _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd
.text:0000000000000224                 add     rsp, 20h
.text:0000000000000228                 lea     rsp, [rbp+0]
.text:000000000000022C                 pop     rbp
.text:000000000000022D                 retn
.text:000000000000022D _D6test644foo4FZv endp

```

**Comment by 9il** [comment]
Nice, foo3 and foo4 looks identical.
Can you past also `_D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd`?

**Comment by DmitryOlshansky** [comment]
```
.text:0000000000000248 _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd proc near
.text:0000000000000248
.text:0000000000000248 var_20          = qword ptr -20h
.text:0000000000000248 var_10          = qword ptr -10h
.text:0000000000000248 var_8           = qword ptr -8
.text:0000000000000248 arg_0           = qword ptr  10h
.text:0000000000000248 arg_8           = qword ptr  18h
.text:0000000000000248
.text:0000000000000248                 push    rbp
.text:0000000000000249                 mov     rbp, rsp
.text:000000000000024C                 sub     rsp, 20h
.text:0000000000000250                 movsd   [rbp+arg_0], xmm0
.text:0000000000000255                 movsd   [rbp+arg_8], xmm1
.text:000000000000025A                 fld     [rbp+arg_8]
.text:000000000000025D                 fabs
.text:000000000000025F                 fstp    [rbp+var_20]
.text:0000000000000262                 movsd   xmm1, [rbp+var_20]
.text:0000000000000267                 movsd   [rbp+var_10], xmm1
.text:000000000000026D                 movsd   [rbp+var_20], xmm1
.text:0000000000000272                 fld     [rbp+var_20]
.text:0000000000000275                 fld     [rbp+arg_0]
.text:0000000000000278                 fabs
.text:000000000000027A                 fstp    [rbp+var_20]
.text:000000000000027D                 movsd   xmm0, [rbp+var_20]
.text:0000000000000282                 movsd   [rbp+var_8], xmm0
.text:0000000000000288                 movsd   [rbp+var_20], xmm0
.text:000000000000028D                 fld     [rbp+var_20]
.text:0000000000000290                 fcomip  st, st(1)
.text:0000000000000292                 fstp    st
.text:0000000000000294                 jp      short loc_298
.text:0000000000000296                 jbe     short loc_2EE
.text:0000000000000298
.text:0000000000000298 loc_298:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+4Cj
.text:0000000000000298                 movsd   xmm2, [rbp+var_10]
.text:000000000000029E                 movsd   [rbp+var_8], xmm2
.text:00000000000002A4                 fld     [rbp+arg_0]
.text:00000000000002A7                 fabs
.text:00000000000002A9                 fstp    [rbp+var_20]
.text:00000000000002AC                 movsd   xmm3, [rbp+var_20]
.text:00000000000002B1                 movsd   [rbp+var_10], xmm3
.text:00000000000002B7                 movsd   [rbp+var_20], xmm3
.text:00000000000002BC                 fld     [rbp+var_20]
.text:00000000000002BF                 fld     cs:_TMP2
.text:00000000000002C5                 fucomip st, st(1)
.text:00000000000002C7                 fstp    st
.text:00000000000002C9                 jp      short loc_2D1
.text:00000000000002CB                 jz      loc_456
.text:00000000000002D1
.text:00000000000002D1 loc_2D1:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+81j
.text:00000000000002D1                 fld     [rbp+var_8]
.text:00000000000002D4                 fld     cs:_TMP2
.text:00000000000002DA                 fucomip st, st(1)
.text:00000000000002DC                 fstp    st
.text:00000000000002DE                 jnz     short loc_2EE
.text:00000000000002E0                 jp      short loc_2EE
.text:00000000000002E2                 movsd   xmm0, [rbp+var_8]
.text:00000000000002E8                 lea     rsp, [rbp+0]
.text:00000000000002EC                 pop     rbp
.text:00000000000002ED                 retn
.text:00000000000002EE ; ---------------------------------------------------------------------------
.text:00000000000002EE
.text:00000000000002EE loc_2EE:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+4Ej
.text:00000000000002EE                                         ; _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+96j ...
.text:00000000000002EE                 fld     [rbp+var_8]
.text:00000000000002F1                 fld     cs:_TMP3
.text:00000000000002F7                 fcomip  st, st(1)
.text:00000000000002F9                 fstp    st
.text:00000000000002FB                 ja      loc_39B
.text:0000000000000301                 jp      loc_39B
.text:0000000000000307                 movsd   xmm2, cs:_TMP4
.text:0000000000000310                 movsd   xmm3, [rbp+var_10]
.text:0000000000000316                 mulsd   xmm3, xmm2
.text:000000000000031A                 movsd   [rbp+var_10], xmm3
.text:0000000000000320                 movsd   xmm4, cs:_TMP4
.text:0000000000000329                 movsd   xmm0, [rbp+var_8]
.text:000000000000032F                 mulsd   xmm0, xmm4
.text:0000000000000333                 movsd   [rbp+var_8], xmm0
.text:0000000000000339                 movsd   xmm1, [rbp+var_10]
.text:000000000000033F                 movsd   xmm2, [rbp+var_10]
.text:0000000000000345                 mulsd   xmm2, xmm1
.text:0000000000000349                 movsd   [rbp+var_10], xmm2
.text:000000000000034F                 movsd   xmm3, [rbp+var_8]
.text:0000000000000355                 movsd   xmm4, [rbp+var_8]
.text:000000000000035B                 mulsd   xmm4, xmm3
.text:000000000000035F                 movsd   [rbp+var_8], xmm4
.text:0000000000000365                 movsd   xmm0, [rbp+var_10]
.text:000000000000036B                 movsd   xmm1, [rbp+var_8]
.text:0000000000000371                 addsd   xmm1, xmm0
.text:0000000000000375                 movsd   [rbp+var_8], xmm1
.text:000000000000037B                 fld     [rbp+var_8]
.text:000000000000037E                 fsqrt
.text:0000000000000380                 fstp    [rbp+var_20]
.text:0000000000000383                 movsd   xmm0, [rbp+var_20]
.text:0000000000000388                 movsd   xmm2, cs:_TMP5
.text:0000000000000391                 mulsd   xmm0, xmm2
.text:0000000000000395                 lea     rsp, [rbp+0]
.text:0000000000000399                 pop     rbp
.text:000000000000039A                 retn
.text:000000000000039B ; ---------------------------------------------------------------------------
.text:000000000000039B
.text:000000000000039B loc_39B:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+B3j
.text:000000000000039B                                         ; _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+B9j
.text:000000000000039B                 fld     [rbp+var_10]
.text:000000000000039E                 fld     cs:_TMP6
.text:00000000000003A4                 fcomip  st, st(1)
.text:00000000000003A6                 fstp    st
.text:00000000000003A8                 jb      loc_442
.text:00000000000003AE                 movsd   xmm1, cs:_TMP7
.text:00000000000003B7                 movsd   xmm2, [rbp+var_10]
.text:00000000000003BD                 mulsd   xmm2, xmm1
.text:00000000000003C1                 movsd   [rbp+var_10], xmm2
.text:00000000000003C7                 movsd   xmm3, cs:_TMP7
.text:00000000000003D0                 movsd   xmm4, [rbp+var_8]
.text:00000000000003D6                 mulsd   xmm4, xmm3
.text:00000000000003DA                 movsd   [rbp+var_8], xmm4
.text:00000000000003E0                 movsd   xmm0, [rbp+var_10]
.text:00000000000003E6                 movsd   xmm1, [rbp+var_10]
.text:00000000000003EC                 mulsd   xmm1, xmm0
.text:00000000000003F0                 movsd   [rbp+var_10], xmm1
.text:00000000000003F6                 movsd   xmm2, [rbp+var_8]
.text:00000000000003FC                 movsd   xmm3, [rbp+var_8]
.text:0000000000000402                 mulsd   xmm3, xmm2
.text:0000000000000406                 movsd   [rbp+var_8], xmm3
.text:000000000000040C                 movsd   xmm4, [rbp+var_10]
.text:0000000000000412                 movsd   xmm0, [rbp+var_8]
.text:0000000000000418                 addsd   xmm0, xmm4
.text:000000000000041C                 movsd   [rbp+var_8], xmm0
.text:0000000000000422                 fld     [rbp+var_8]
.text:0000000000000425                 fsqrt
.text:0000000000000427                 fstp    [rbp+var_20]
.text:000000000000042A                 movsd   xmm0, [rbp+var_20]
.text:000000000000042F                 movsd   xmm1, cs:_TMP8
.text:0000000000000438                 mulsd   xmm0, xmm1
.text:000000000000043C                 lea     rsp, [rbp+0]
.text:0000000000000440                 pop     rbp
.text:0000000000000441                 retn
.text:0000000000000442 ; ---------------------------------------------------------------------------
.text:0000000000000442
.text:0000000000000442 loc_442:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+160j
.text:0000000000000442                 fld     [rbp+var_10]
.text:0000000000000445                 fmul    cs:_TMP9
.text:000000000000044B                 fld     [rbp+var_8]
.text:000000000000044E                 fcomip  st, st(1)
.text:0000000000000450                 fstp    st
.text:0000000000000452                 jnb     short loc_462
.text:0000000000000454                 jp      short loc_462
.text:0000000000000456
.text:0000000000000456 loc_456:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+83j
.text:0000000000000456                 movsd   xmm0, [rbp+var_10]
.text:000000000000045C                 lea     rsp, [rbp+0]
.text:0000000000000460                 pop     rbp
.text:0000000000000461                 retn
.text:0000000000000462 ; ---------------------------------------------------------------------------
.text:0000000000000462
.text:0000000000000462 loc_462:                                ; CODE XREF: _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+20Aj
.text:0000000000000462                                         ; _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd+20Cj
.text:0000000000000462                 movsd   xmm4, [rbp+var_10]
.text:0000000000000468                 movsd   xmm1, [rbp+var_10]
.text:000000000000046E                 mulsd   xmm1, xmm4
.text:0000000000000472                 movsd   [rbp+var_10], xmm1
.text:0000000000000478                 movsd   xmm0, [rbp+var_8]
.text:000000000000047E                 movsd   xmm2, [rbp+var_8]
.text:0000000000000484                 mulsd   xmm2, xmm0
.text:0000000000000488                 movsd   [rbp+var_8], xmm2
.text:000000000000048E                 movsd   xmm3, [rbp+var_10]
.text:0000000000000494                 movsd   xmm4, [rbp+var_8]
.text:000000000000049A                 addsd   xmm4, xmm3
.text:000000000000049E                 movsd   [rbp+var_8], xmm4
.text:00000000000004A4                 fld     [rbp+var_8]
.text:00000000000004A7                 fsqrt
.text:00000000000004A9                 fstp    [rbp+var_20]
.text:00000000000004AC                 movsd   xmm0, [rbp+var_20]
.text:00000000000004B1                 lea     rsp, [rbp+0]
.text:00000000000004B5                 pop     rbp
.text:00000000000004B6                 retn
.text:00000000000004B6 _D3std4math14__T5hypotTdTdZ5hypotFNaNbNiNfxdxdZd endp
```

**Comment by DmitryOlshansky** [comment]
> Nice, foo3 and foo4 looks identical.

OMG. Seems like abs version calls the wrong function. And it's caused by `-inline` switch. I tried all combinations of -O and -inline, -O has no negative effect.

**Comment by 9il** [comment]
>> Nice, foo3 and foo4 looks identical.

> OMG. Seems like abs version calls the wrong function. And it's caused by -inline switch. I tried all combinations of -O and -inline, -O has no negative effect.

Why you think so? Should not `abs` calls `hypot` for complex numbers?

**Comment by DmitryOlshansky** [comment]
> Why you think so? Should not abs calls hypot for complex numbers?

Ehm right, I must be tired.

**Comment by 9il** [comment]
Hypot code looks really crazy. Excuse me for please for a lot of asm requests. This is another one:
```D
void foo5(){
    cdouble c = -1+1i;
    auto a = hypot(c.re, cast(double)c.im); //idouble -> double?
}
```

**Comment by 9il** [comment]
foo5 is fine, you can skip previous msg.

**Comment by 9il** [comment]
I can not find wrong code. 
Can you please upload code for all file compiled with `-w -dip25 -m32 -fPIC -O -release -unittest -c` (like in auto-tester)?

**Comment by DmitryOlshansky** [comment]
64bit binary:
https://www.dropbox.com/s/plwv9ce3zs9vfk0/math_64.exe?dl=1

32bit binary (also asserts for me):
https://www.dropbox.com/s/ec9jwxjoap1k7n8/math.exe?dl=1

Both with debug info I hope you can navigate in them with disassembler.

**Comment by 9il** [comment]
otool -tvV ./math.exe 
./math.exe: is not an object file

Please send the text file.

**Comment by DmitryOlshansky** [comment]
https://www.dropbox.com/s/4xrzbmllef6bjtv/asm.zip?dl=1

**Comment by 9il** [comment]
I can not find any of `fooX` because function names defined like `sub_402DF0 proc	near` :(

**Comment by DmitryOlshansky** [comment]
I'm certain 64-bit version has names. 32-bit sadly has broken debug symbols and disassembler fails to load them.

**Comment by 9il** [comment]
I have 32 bit version now )

**Comment by 9il** [comment]
But I can not catch error on Mac, because Darwin_64_32 not fails. So my 32bit binaries are useless.

**Comment by DmitryOlshansky** [comment]
Is there VirtualBox for Mac? At home my Linux sits in VBox on Windows host and find it incredibly easy to switch between the two.

**Comment by 9il** [comment]
I‘ll try with VB.

**Comment by greenify** [comment]
Merge conflict - please rebase :)

**Pull request closed without merging**